avr-hal-generic/src/i2c.rs

```diff
@@ -329,51 +336,110 @@ macro_rules! impl_twi_i2c {
+                //    Register (TWDR) must be complete before clearing this flag.
                 //
                 //
                 //    Bit 5 – TWSTA: TWI START Condition Bit
                 //
-                //    The application writes the TWSTA bit to one when it desires to become a Master on the 2-wire Serial Bus. The
-                //    TWI hardware checks if the bus is available, and generates a START condition on the bus if it is free. However,
-                //    if the bus is not free, the TWI waits until a STOP condition is detected, and then generates a new START
-                //    condition to claim the bus Master status. TWSTA must be cleared by software when the START condition has
-                //    been transmitted.
+                //    The application writes the TWSTA bit to one when it desires to become a
+                //    Master on the 2-wire Serial Bus. The TWI hardware checks if the bus is
+                //    available, and generates a START condition on the bus if it is free. However,
+                //    if the bus is not free, the TWI waits until a STOP condition is detected, and
+                //    then generates a new START condition to claim the bus Master status. TWSTA
+                //    must be cleared by software when the START condition has been transmitted.
                 //
                 //    NOTE: `twstart` macro token is defined as `twsta` in chips/[CHIP]/src/lib.rs
                 self.p.$twcr.write(|w| w
                     .$twen().set_bit()
                     .$twint().set_bit()
                     .$twstart().set_bit()
                 );
-                // NOTE: why do we need to wait here? `wait()` is just reading twcr.twint to see if
-                // the bit is set...but it was just set in the line above...
+
+                // QUESTION: why do we need to wait here? `wait()` is just reading twcr.twint to see if
+                // the bit is set...but it was just set in the line above... Is there some sort of
+                // timing issue here that is being papered over by the general wait()?
+                // I can see that below the TWDR is writeable when the TWINT is set (by hardware it
+                // says):
+                //
+                // (datsheet) 22.9.2 TWCR – TWI Control Register
+                // (datasheet) 22.9.4 TWDR – TWI Data Register Bit
+                //
+                //    TWDR In Transmit mode, TWDR contains the next byte to be transmitted. In
+                //    Receive mode, the TWDR contains the last byte received. It is writable while
+                //    the TWI is not in the process of shifting a byte. This occurs when the TWI
+                //    Interrupt Flag (TWINT) is set by hardware.
                 self.wait();
 
                 // Validate status
+                //
+                // NOTE: here's what this impl looks like in lpc8xx-hal, modeled as a state
```

Comment:
I like the class state machine pattern in rust. Just showing what lpc8xx does here. Perhaps we can adopt this pattern?